src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests5.cs

```diff
@@ -0,0 +1,236 @@
+                //         _ = new C() is { Prop1: null, Prop1.Prop2: null };
+                Diagnostic(ErrorCode.ERR_IsPatternImpossible, "new C() is { Prop1: null, Prop1.Prop2: null }").WithArguments("C").WithLocation(10, 13)
+                );
+        }
+
+        [Fact]
+        public void ExtendedPropertyPatterns_03()
+        {
+            var program = @"
+using System;
+class C
+{
+    public C Prop1 { get; set; }
+    public C Prop2 { get; set; }
+
+    public static void Main(string[] args)
+    {        
+        var result = new C() switch
+        {
+            {Prop1: null} => 1,
+            {Prop1.Prop2: null} => 2,
+            _ => -1,
+        };
+        Console.Write(result);
+    }
+}
+";
+            var compilation = CreateCompilation(program, parseOptions: TestOptions.RegularWithExtendPropertyPattern, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics();
+            var verifier = CompileAndVerify(compilation);
+            verifier.VerifyIL("C.Main", @"
+{
+  // Code size       46 (0x2e)
+  .maxstack  1
+  .locals init (int V_0,
+                C V_1,
+                C V_2)
+  IL_0000:  newobj     ""C..ctor()""
+  IL_0005:  stloc.1
+  IL_0006:  ldloc.1
+  IL_0007:  brfalse.s  IL_0025
+  IL_0009:  ldloc.1
+  IL_000a:  callvirt   ""C C.Prop1.get""
+  IL_000f:  stloc.2
+  IL_0010:  ldloc.2
+  IL_0011:  brfalse.s  IL_001d
+  IL_0013:  ldloc.2
+  IL_0014:  callvirt   ""C C.Prop2.get""
+  IL_0019:  brfalse.s  IL_0021
+  IL_001b:  br.s       IL_0025
+  IL_001d:  ldc.i4.1
+  IL_001e:  stloc.0
+  IL_001f:  br.s       IL_0027
+  IL_0021:  ldc.i4.2
+  IL_0022:  stloc.0
+  IL_0023:  br.s       IL_0027
+  IL_0025:  ldc.i4.m1
+  IL_0026:  stloc.0
+  IL_0027:  ldloc.0
+  IL_0028:  call       ""void System.Console.Write(int)""
+  IL_002d:  ret
+}
+");
+        }
+
+        [Fact]
+        public void ExtendedPropertyPatterns_04()
+        {
+            var program = @"
+class C
+{
+    public C Prop1 { get; set; }
+    public C Prop2 { get; set; }
+
+    public static void Main(string[] args)
+    {        
+        _ = new C() is { Prop1?.Prop2: {} };
+    }
+}
+";
+            var compilation = CreateCompilation(program, parseOptions: TestOptions.RegularWithExtendPropertyPattern, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics(
+                // (9,26): error CS9001: Conditional access may not be used in property subpatterns. Use a simple member access instead.
+                //         _ = new C() is { Prop1?.Prop2: {} };
+                Diagnostic(ErrorCode.ERR_ConditionalAccessInSubpattern, "Prop1?.Prop2").WithLocation(9, 26));
+        }
+
+        [Fact]
+        public void ExtendedPropertyPatterns_05()
+        {
+            var program = @"
+class C
+{
+    public static void Main(string[] args)
+    {        
+        _ = new C() is { Prop1<int>.Prop2: {} };
```

Comment:
It would be useful to test various expressions that are valid patterns but aren't `MemberAccessExpressionSyntax` or `IdentifierNameSyntax` when converted to expressions, and so should error out. 
`{ [0]: { } }` is interesting for example, because `[0]` will be a length-pattern soon.
Other patterns currently supported in this branch may be useful too: `{ not 1: { } }`, `{ 1: { } }`, `{ > 1: { } }`, ...

Reply:
>`{ [0]: { } }` 

That'll be easy to cover. If get a length pattern and there's a colon we convert to MemberBindingExpression. 
Right now I think https://github.com/dotnet/roslyn/pull/52139#discussion_r619675420 is more of an issue if we want to give meaningful diagnostics.